lntypes/preimage.go

```diff
@@ -9,28 +9,40 @@ import (
 // PreimageSize of array used to store preimagees.
 const PreimageSize = 32
 
-// Preimage is used in several of the lightning messages and common structures. It
-// represents a payment preimage.
+// Preimage is used in several of the lightning messages and common structures.
+// It represents a payment preimage.
 type Preimage [PreimageSize]byte
 
 // String returns the Preimage as a hexadecimal string.
 func (p Preimage) String() string {
 	return hex.EncodeToString(p[:])
 }
 
-// NewPreimage returns a new Preimage from a byte slice.  An error is returned if
-// the number of bytes passed in is not PreimageSize.
+// NewPreimage allocates a new Preimage from a byte slice. An error is returned
+// if the number of bytes passed in is not PreimageSize.
 func NewPreimage(newPreimage []byte) (*Preimage, error) {
+	preimage, err := MakePreimage(newPreimage)
+	if err != nil {
+		return nil, err
+	}
+
+	return &preimage, nil
+}
+
+// MakePreimage returns a new Preimage from a bytes slice. An error is returned
+// if the number of bytes passed in is not PreimageSize.
+func MakePreimage(newPreimage []byte) (Preimage, error) {
```

Comment:
Yes I also ran into that ref return from `NewPreimage`. In the hodl invoice branch, I chose to just change the return type to a value. I couldn't see a good use case for the ptr return.

Reply:
sounds good, i'm okay with either approach. indeed i found i didn't need the pointer object in this PR, but i could imagine cases where it would be useful

Reply:
I think just one constructor is enough, but it is all not very important.

Reply:
Would prefer sticking to one constructor to avoid confusion :)

Reply:
In any case, please keep the non-pointer named `Make`. In another PR I just went through a rebase action to do that. I understand that there is a convention where `New` is for pointers and `Make` for values.

Reply:
My personal preference is typically to add both constructors, which allows the caller to use whatever is best as requiring an extra indirect or dereference can have performance impacts or make calling code more verbose. It seems like we don't really use the pointer version tho atm, so i just decided to leave `MakePreimage` that returns the non-pointer version. We can always add `NewPreimage` later if we want/need to